File: common/src/main/java/org/valkyrienskies/clockwork/content/contraptions/flap/FlapBearingBlock.java
package org.valkyrienskies.clockwork.content.contraptions.flap;

import com.simibubi.create.content.contraptions.bearing.BearingBlock;
import com.simibubi.create.foundation.block.IBE;
import net.minecraft.core.BlockPos;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.context.UseOnContext;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.entity.BlockEntityType;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.block.state.StateDefinition;
import net.minecraft.world.phys.BlockHitResult;
import org.valkyrienskies.clockwork.ClockWorkBlockEntities;

public class FlapBearingBlock extends BearingBlock implements IBE<FlapBearingBlockEntity> {
    public FlapBearingBlock(Properties properties) {
        super(properties);
    }

    @Override
    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> builder) {
        super.createBlockStateDefinition(builder);
    }

    @Override
    public Class<FlapBearingBlockEntity> getBlockEntityClass() {
        return FlapBearingBlockEntity.class;
    }

    @Override
    public InteractionResult use(BlockState state, Level worldIn, BlockPos pos,
                                 Player player, InteractionHand handIn, BlockHitResult hit) {
        if (!player.mayBuild())
            return InteractionResult.FAIL;
        if (player.isShiftKeyDown())
            return InteractionResult.FAIL;
        if (player.getItemInHand(handIn).isEmpty()) {
            if (!worldIn.isClientSide) {
                withBlockEntityDo(worldIn, pos, te -> {
                    if (te.running) {
                        te.disassemble();
                        return;
                    }
                    te.assembleNextTick = true;
                });
            }
            return InteractionResult.SUCCESS;
        }
        return InteractionResult.PASS;
    }

    @Override
    public BlockEntityType<? extends FlapBearingBlockEntity> getBlockEntityType() {
        return ClockWorkBlockEntities.FLAP_BEARING.get();
    }

    @Override
    public InteractionResult onWrenched(BlockState state, UseOnContext context) {
        InteractionResult resultType = super.onWrenched(state, context);
        if (!context.getLevel().isClientSide && resultType.consumesAction())
            withBlockEntityDo(context.getLevel(), context.getClickedPos(), FlapBearingBlockEntity::disassemble);
        return resultType;
    }
}

File: common/src/main/kotlin/org/valkyrienskies/clockwork/content/contraptions/flap/FlapBearingBlock.kt
package org.valkyrienskies.clockwork.content.contraptions.flap

import com.simibubi.create.content.contraptions.bearing.BearingBlock
import com.simibubi.create.foundation.block.IBE
import net.minecraft.core.BlockPos
import net.minecraft.world.InteractionHand
import net.minecraft.world.InteractionResult
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.context.UseOnContext
import net.minecraft.world.level.Level
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraft.world.level.block.entity.BlockEntityType
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.level.block.state.StateDefinition
import net.minecraft.world.phys.BlockHitResult
import org.valkyrienskies.clockwork.ClockworkBlockEntities
import java.util.function.Consumer

class FlapBearingBlock(properties: Properties?) :
    BearingBlock(properties),
    IBE<FlapBearingBlockEntity> {
    override fun createBlockStateDefinition(builder: StateDefinition.Builder<Block, BlockState>) {
        super.createBlockStateDefinition(builder)
    }

    override fun getBlockEntityClass(): Class<FlapBearingBlockEntity> {
        return FlapBearingBlockEntity::class.java
    }

    override fun use(
        state: BlockState,
        worldIn: Level,
        pos: BlockPos,
        player: Player,
        handIn: InteractionHand,
        hit: BlockHitResult
    ): InteractionResult {
        if (!player.mayBuild()) return InteractionResult.FAIL
        if (player.isShiftKeyDown) return InteractionResult.FAIL
        if (player.getItemInHand(handIn).isEmpty) {
            if (!worldIn.isClientSide) {
                withBlockEntityDo(
                    worldIn,
                    pos,
                    Consumer withBlockEntityDo@{ te: FlapBearingBlockEntity ->
                        if (te.isRunning) {
                            te.disassemble()
                            return@withBlockEntityDo
                        }
                        te.assembleNextTick = true
                    }
                )
            }
            return InteractionResult.SUCCESS
        }
        return InteractionResult.PASS
    }

    override fun getBlockEntityType(): BlockEntityType<out FlapBearingBlockEntity> {
        return ClockworkBlockEntities.FLAP_BEARING.get()
    }

    override fun newBlockEntity(p_153215_: BlockPos, p_153216_: BlockState): BlockEntity {
        val isSmart = this.descriptionId == "block.vs_clockwork.smart_flap_bearing"
        val maxSize = if (isSmart) -1L else 16L
        return FlapBearingBlockEntity(blockEntityType, p_153215_, p_153216_, maxSize, isSmart)
    }

    override fun onWrenched(state: BlockState, context: UseOnContext): InteractionResult {
        val resultType = super.onWrenched(state, context)
        if (!context.level.isClientSide && resultType.consumesAction()) withBlockEntityDo(
            context.level,
            context.clickedPos,
            FlapBearingBlockEntity::disassemble
        )
        return resultType
    }
}
